### PR TITLE
[None][fix] Clamp very small non-zero temperature in SamplingParams

### LINKED ISSUE #15715: [Bug]: Clamp very small non-zero temperature values to avoid numerical instability
### System Info

CPU architecture: x86_64

CPU: Intel Xeon Gold 6462C, 8 vCPUs (4 cores × 2 threads), max 3900 MHz

GPU name: NVIDIA L20
GPU memory size: 48G

TensorRT-LLM branch or tag: main
TensorRT-LLM commit: https://github.com/NVIDIA/TensorRT-LLM/commit/28acbad14b442e89312cfe9e845a5f721a3e702f
PyTorch: 2.10.0+cu130
CUDA: 13.0.88
cuDNN: 9.15.0
NVIDIA driver version: 550.163.01

OS: Ubuntu 24.04.3 LTS

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

```python
from tensorrt_llm.sampling_params import SamplingParams

# Before the fix: very small non-zero temperature bypasses validation
# and enters the sampling backend, causing potential inf/nan in logits/temperature
sp = SamplingParams(temperature=1e-12, top_k=50)
print(sp.temperature)  # prints 1e-12 — dangerously small

# Simulating what happens inside the backend:
import torch
logits = torch.tensor([1.0, 2.0, 3.0])
temperature = 1e-12
scaled = logits / temperature
print(scaled)  # tensor([1.0e+12, 2.0e+12, 3.0e+12]) → leads to inf/nan in softmax
print(torch.softmax(scaled, dim=-1))  # tensor([0., 0., 1.]) or nan depending on dtype
```

### Expected behavior

SamplingParams should reject or safely normalize temperature values that are 
positive but below a numerical stability threshold.

Specifically:
- temperature=0.0  → kept as 0.0 (greedy decoding)
- temperature=1e-12 → clamped to MIN_SAMPLING_TEMPERATURE (1e-2)
- temperature=1.0  → kept as 1.0 (normal sampling)

### actual behavior

amplingParams only validates temperature >= 0, so values like 1e-12 
pass validation unchanged and are forwarded to the sampling backend.

Inside the backend, logits / temperature with temperature=1e-12 produces 
extremely large logit values (e.g., 1e+12), which overflow to inf or nan 
in float16/bfloat16 precision, leading to undefined sampling behavior.

Example:
```python
import torch
logits = torch.tensor([1.0, 2.0, 3.0], dtype=torch.float16)
scaled = logits / 1e-12
print(scaled)
# tensor([inf, inf, inf], dtype=torch.float16)

print(torch.softmax(scaled, dim=-1))
# tensor([nan, nan, nan], dtype=torch.float16)
```

### additional notes

1.  This fix mirrors the same numerical stability guard used by vLLM    (temperature = max(temperature, 1e-2) for non-zero values).

2. The issue is most severe in float16/bfloat16: typical logit values (~10)    divided by 1e-6 already exceed float16's max range (~65504),    producing inf/nan in softmax.

3. temperature=0.0 (greedy decoding) is intentionally excluded from the clamp    and its behavior is unchanged.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by CodersAcademy006**
## Description

Fixes #15715.

`SamplingParams` only validated `temperature >= 0`, so a positive but tiny value like `1e-12` passed validation unchanged and was forwarded to the sampling backend. There, `logits / temperature` overflows to `inf`/`nan` in `float16`/`bfloat16` (a typical logit ~10 divided by `1e-12` is ~`1e13`, far past fp16's ~65504 max), producing undefined sampling behavior.

This clamps `0 < temperature < MIN_SAMPLING_TEMPERATURE` (`1e-2`) up to that floor, mirroring the guard vLLM uses. Behavior for the other cases is unchanged:

| input | result |
|-------|--------|
| `temperature = 0.0` | kept `0.0` (greedy) |
| `0 < temperature < 1e-2` | clamped to `1e-2` (with a warning) |
| `temperature >= 1e-2` | unchanged |
| `temperature < 0` | still rejected |

## Test Coverage

Adds `tests/unittest/llmapi/test_sampling_params_temperature.py` covering the clamp, the greedy (`0.0`) passthrough, the boundary/normal passthrough, and the negative-rejection path.

## PR Checklist
- [x] PR title follows `[JIRA/NVBugs/GitHub issue/None][type] Summary` format
- [x] Test cases added
- [x] Commit is signed off (DCO)

cc @chfeng-cs — thanks for the detailed report and repro.

## Summary by CodeRabbit

* **Bug Fixes**
  * Very small positive temperature values are now automatically raised to a safe minimum, improving sampling stability.
  * Zero temperature remains unchanged for greedy sampling.
  * Negative temperature values continue to be rejected.

* **Tests**
  * Added coverage for minimum, zero, normal, and invalid temperature values.

### CODE REVIEW

**Comment by CodersAcademy006** [comment]
Noting there are two earlier PRs for this same issue — #15716 (from the reporter @chfeng-cs) and #16222 (@lonexreb) — so maintainers can pick whichever they prefer. This one aims to be the most complete of the three:

- clamps `0 < temperature < 1e-2` up to the floor (mirrors vLLM), leaving `temperature = 0` greedy;
- additionally **rejects non-finite temperature** (`inf`/`nan`) up front — the old `>= 0` check let those through (`nan < 0` is `False`, `inf` is unclamped), and they also produce undefined sampling;
- documents the behavior on the `temperature` docstring;
- uses `logger.debug` (not `warning`) to avoid per-request log spam on the serving hot path;
- adds a parametrized test matrix covering unset / greedy / tiny / just-below-floor / boundary / normal, plus negative and non-finite rejection.

Happy to consolidate into one of the existing PRs instead if that's cleaner for review.

**Review by lori-ren — APPROVED** [review]

**Comment by chfeng-cs** [comment]
Thanks for acknowledging #15716.

#15716 was opened on June 29 as the original implementation for #15715 and has been ready for review since its only automated suggestion was addressed that day. It already includes the temperature floor, preserves `temperature=0`, uses `logger.debug`, documents the behavior, and adds parametrized tests. I have also added the useful non-finite validation from this PR.

#15716 keeps the change smaller and more cohesive by extending the existing validation and test suite. Its compact test matrix verifies both the Python value and the value forwarded to the backend through `_get_sampling_config()`, making the change easier to review and maintain.

As an independent contributor, receiving fair attribution and review for original work is important to me.

@lori-ren, could we consolidate the work in the original PR and review #15716 before merging a later duplicate?

**Comment by CodersAcademy006** [comment]
Fair, and you've now folded in the non-finite check too, so there's nothing left here #15716 doesn't already cover. Closing this one in favor of yours — thanks for the thorough test matrix.

**Comment by CodersAcademy006** [comment]
Closing in favor of #15716 per above.
